### ISSUE TITLE
Strict mocks support

### ISSUE THREAD

**Issue by szczepiq** [mockito/mockito#1097]
## Introduction

**Supporting "strict" mocks will help with test debuggability in certain scenarios.**

This design note explains the rationale and suggested implementation plan to support "strict" mocks. **Please contribute!** If you want to help our community and implement this feature, please read the design spec carefully and let us know if something is unclear. This feature has been asked for many years now and we think that now we are ready to nail it. We already started to make Mockito stricter (#769).

We believe that this design note is pretty awesome and you could get started on implementing this feature in no time. Seriously. Read it, code it and give back a pull request :) If something is unclear, you have feedback about the design or better ideas - leave us a comment.

## Rationale

Mockito mocks are loose by default - methods can be freely invoked on a mock, non-void methods will return default empty value. "Strict" mock works differently, you are not allowed to invoke a method on a strict mock unless this method was previously "configured". This is how traditional mocking frameworks like JMock or EasyMock work by default. Strict mocks offer better debuggability in certain scenarios, however overusing strict mocks leads to overspecified, brittle tests. Strict VS loose mock explanation is further described in [an article on Mockito blog](http://blog.mockito.org/2017/01/clean-tests-produce-clean-code-strict.html).

**Mockito mocks will remain loose by default but we want to offer better support for strict mocks.**

Useful links:

- previous design and discussion in #649
- initial discussion in #303

### New behavior

Strict mocking can be enabled per test method execution and introduces following behavior:
 - unstubbed, non-void method fails immediately with "I am unstubbed!" exception
 - such method is implicitly marked as verified for the verifyNoMoreInteractions() calls
 - unused stubs are detected at the end of test and trigger failure
 - requires to use doReturn API for stubbing mocks (limitation of the syntax, already known gotcha when using [test "spy"](https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/Mockito.html#13) or [strict stubs](https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/exceptions/misusing/PotentialStubbingProblem.html))

### Options and questions

 - Q: Do we separate behavior for voids / non-voids? A: Mockito API makes clear distinction between stubbing (typically non-void methods) and verification (typically void methods). It is a design choice to drive clarity and understanding of ASK vs TELL style interactions (this design spec does not explain this topic further). Unless we hear more feedback, we want the strict mock to only fail early on non-void methods.
 - Q: Do we offer Strictness level configuration per mock object? A: Is it needed? So far it was not reported by users that it is needed so we will ignore this use case.

### New API

- new class: ```MockitoJUnitRunner.StrictMocks```, similar to [MockitoJUnitRunner.StrictStubs](https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/junit/MockitoJUnitRunner.java#L142)
- new enum type: ```Strictness.STRICT_MOCKS```, similar to [Strictness.STRICT_STUBS](https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/quality/Strictness.java#L91)

### New API side by side with existing API

MockitoJUnitRunner.Silent
MockitoJUnitRunner.Strict
MockitoJUnitRunner.StrictStubs
**MockitoJUnitRunner.StrictMocks**

MockitoRule.silent()
MockitoRule.strictness(Strictness)

Strictness.LENIENT
Strictness.WARN
Strictness.STRICT_STUBS
**Strictness.STRICT_MOCKS**

### Getting started

Suggested starting point for the implementation:

- unit test, see test that covers similar functionality - [StrictStubbingTest.java](https://github.com/mockito/mockito/blob/release/2.x/src/test/java/org/mockitousage/stubbing/StrictStubbingTest.java). New test needs to demonstrate that exception is thrown when we invoke non-void, unstubbed method on a mock.
- to implement the feature, I think we need to tweak a little bit [DefaultStubbingLookupListener](https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/internal/junit/DefaultStubbingLookupListener.java#L33).

### Test coverage

Suggested test coverage:

- demonstrate the feature works with:
    - MockitoSession API
    - MockitoJUnitRunner
    - MockitoRule
- feature coverage
    - void and non-void methods
    - configure strict mocks for single test method only, all other test methods can have different Strictness
    - detects unused stubs
    - implicitly verified for verifyNoMoreInteractions()
    - documented the need for using doReturn API
        - in Javadoc
        - strict mock verification exception (~“I am unstubbed!”) contains hint
    - if method was stubbed but with different args, the verification exception message should indicate it

**Comment by imod** [comment]
I think it would be very beneficial to have a system property to switch Mockito to strict mode on global level. This would allow enforcing strict mode via CI (e.g. Maven) on a whole project.

**Comment by szczepiq** [comment]
@imod, thank you for feedback! I haven't thought about the idea to use system property.

>...assuming I don't have to use different APIs for strict and not strict mode.

That's correct, the the api for using mocks is the same. The difference lies in how [JUnitRule](https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/junit/MockitoRule.html) or the [MockitoSession](https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/MockitoSession.html) objects are configured.

>I think it would be very beneficial to have a system property to switch Mockito to strict mode on global level. This would allow enforcing strict mode via CI (e.g. Maven) on a whole project.

My comments:

- This could be implemented without making changes in the framework - base class for tests that uses MockitoSession or junit rule, configure the latter via System property.
- I'm skeptical about making the CI run tests with mocks differently then during local development. This could be very confusing.

**Comment by imod** [comment]
@szczepiq I agree, it could be confusing - but to be honest: I think this is a decision not a library has to make. If you are responsible for the development of a product, then I think it is not uncommon to define such rules and this for sure would be one I would set - in fact we don't use mockito because of the lack of it.

**Comment by szczepiq** [comment]
@imod, thank you for feedback!

This ticket is to add support for strict mocks. It does not offer globally configurable "strictness level". We will see if there is more feedback from the community to add this sort of 'global' setting. At the moment I'd rather not create implicit ways of configuring default behavior of the framework.

Ok, who wants to work on this feature? The design spec is ready to implement! :)

**Comment by TehBakker** [comment]
Just to make sure, there is currently no way to define that for a class / test I want my mock to all act as strict and throw an error if a call was not declared ?

**Comment by piffta** [comment]
I completely agree with @imod. That is only reason I am avoiding to use Mockito. 
Adding strict mode will make debugging easier, for example if test have to be updated after some implementation changes.

**Comment by jadeleeuw** [comment]
I am currently toying around with this feature to see if I could implement it, but there is one thing I am struggling with so far. I am not very familiar yet with the internals of Mockito, so please excuse my ignorance.

Every time a method is invoked on a mock, a `StubbingLookupEvent` is created and processed in `onStubbingLookup` of [DefaultStubbingLookupListener](https://github.com/mockito/mockito/blob/ab569c0ca4db1d0ffd98ea5251f08c825b5b4c19/src/main/java/org/mockito/internal/junit/DefaultStubbingLookupListener.java#L37). As highlighted in the `Getting started` section of this issue, this is most likely also where the logic for this feature should reside.

However since this `StubbingLookupEvent` is created every time a method is invoked, this means that this event is also created during the creation of a stub. Which makes sense, since stubs can be overwritten.

Which leads where I am stuck: How do I differentiate between a method invocation while creating a stub (e.g. `when(mock.foo(1)).thenReturn("bar")`) and  a regular method invocation used during a test (e.g. `mock.foo(1)`)?

Without this distinction I am unsure how to cause the following specified behaviour:
   
- `unstubbed, non-void method fails immediately with "I am unstubbed!" exception`

Because if this distinction between creation of a stub and (potential) usage of a stub cannot be made and the specified behaviour is implemented, it will always throw the exception during the creation of a non-void stub, making it impossible to create non-void stubs. (because as explained above, creation of a stub also triggers a `StubbingLookupEvent`).

**Comment by TimvdLippe** [comment]
@jadeleeuw Sadly this is a fundamental issue that we hit before and it is currently not possible to distinguish the two. This is also the reason I made a prototype to circumvent this fundamental issue in #1264.

The only way we can patch it right now is not immediately throw, but instead collect all information in the end and evaluate it at once. I think that is what @sambarker was aiming to solve in #1629, but for JUnit5 instead. Right now there is a discrepancy between JUnit4 and JUnit5 implementation, which we need to resolve one way or the other.

**Comment by SamBarker** [comment]
My motivation was to support having a stub definition provided in a `@Before` to only be invoked from a subset of methods rather than all. 

Having said that @timvdlippe is right about what my implementation is doing, by collecting all invocations and evaluating them in `@AfterClass`.

**Comment by mockitoguy** [comment]
@jadeleeuw, it's a limitation of our syntax. In the description, we stated that strict mocking will require using doReturn API.

@henri-tremblay suggested an interesting idea, new method "setStrict(mock)" in https://github.com/mockito/mockito/issues/1677#issuecomment-479013893

**Comment by mockitoguy** [comment]
FYI: I'm making progress on the implementation. Stay tuned :-)

**Comment by zdanek** [comment]
Tuned tuned but nothing happened...

**Comment by Marcono1234** [comment]
> Strict mocks offer better debuggability in certain scenarios, however overusing strict mocks leads to overspecified, brittle tests.

Not using strict mocks can also lead to overspecified tests when you start using `verify(...)` and `verifyNoMoreInteractions(...)` and then end up being too specific with `times(...)`. For example verifying that a getter method was called a specific amount of times, even though in the end the exact number of calls is an implementation detail.
The issue with `verifyNoMoreInteractions(...)` is also that it is easy to forget to call it with some mock, and another issue is that you only call it in the end, and by that time you might have encountered already a `NullPointerException` or an assertion error due to an unstubbed method having been called, and behaving unexpectedly.

> - such method is implicitly marked as verified for the verifyNoMoreInteractions() calls
> - unused stubs are detected at the end of test and trigger failure

But detection of unstubbed methods will not require usage of `verifyNoMoreInteractions(...)`, right? The sentence here only describes how it would behave if both are used together. Maybe that should be clarified.

> Unless we hear more feedback, we want the strict mock to only fail early on non-void methods.

`void` methods are usually expected to have some side-effect, so it is helpful to notice when such a method is called unexpectedly. For example, if you mock a class implementing `Closeable`, and the tested code unexpectedly calls `close()`, then you would like the test to fail (instead of silently passing). Because in production that unexpected `close()` call will most likely break the application if something else afterwards assumes the resource has not been closed.

> MockitoJUnitRunner.Silent
> MockitoJUnitRunner.Strict
> MockitoJUnitRunner.StrictStubs
> **MockitoJUnitRunner.StrictMocks**

So `StrictMocks` would be an extension of `StrictSubs`? Or if not, would it not be possible to have both active at the same time then?

Also, could this configuration maybe be implemented in an easier way than having all these separate runner classes? Could there for example be a `@MockitoRunnerSettings` annotation on the test class which the MockitoJUnitRunner reads (not sure if this is possible)? E.g.:
```java
@RunWith(MockitoJUnitRunner.class)
@MockitoRunnerSettings(failOnUnstubbedUsage = true, failOnUnusedStub = true)
class MyTest {
    ...
}
```

----

What might also be useful is if not only Mockito directly threw an exception when an unstubbed method is called, but also recorded this in the Mockito Session, and checked for it after the test. So that even in case the user code accidentally discarded the Mockito exception (for example only logged it), the test would still fail in the end.

----

Generally I would also love if Mockito was more strict by default, or if it would at least be easier to configure this strictness for a complete test class (or globally) instead of per mock with a default throwing answer.

Currently it can happen that you accidentally write unrealistic tests due to using unstubbed methods, whose behavior would not be possible like this in production.
For example if you forget to stub a `@NonNull` method, and then are accidentally testing how the caller reacts to `null`. Or for example if you stub `List#get` but not `size()` and then by accident have a non-empty `List` whose size is allegedly 0. And in the worst case you then start adjusting production code to account for these unrealistic cases.

(Some of these issues are made up, but I think they are quite realistic.)
